Lombiq.HelpfulLibraries/Libraries/Utilities/StringExtensions.cs

```diff
@@ -52,8 +52,38 @@ public static class StringExtensions
         /// <summary>
         /// A shortcut for <c>string.Replace(string, string, StringComparison.OrdinalIgnoreCase)</c>.
         /// </summary>
-        public static string ReplaceOrdinalIgnoreCase(this string text, string oldValue, string? newValue) =>
-            text.Replace(oldValue, newValue, StringComparison.OrdinalIgnoreCase);
+        public static string ReplaceOrdinalIgnoreCase(this string text, string oldValue, string? newValue = "") =>
+            text.Replace(oldValue, newValue ?? string.Empty, StringComparison.OrdinalIgnoreCase);
+
+        /// <summary>
+        /// A shortcut for <c>string.Equals(string, StringComparison.Ordinal)</c>.
+        /// </summary>
+        public static bool EqualsOrdinal(this string text, string? value) =>
+            text.Equals(value, StringComparison.Ordinal);
+
+        /// <summary>
+        /// A shortcut for <c>string.Contains(string, StringComparison.Ordinal)</c>.
+        /// </summary>
+        public static bool ContainsOrdinal(this string text, string value) =>

```

Comment:
Looking at [the docs](https://docs.microsoft.com/en-us/dotnet/api/system.string.contains?view=net-6.0) for `.Contains()`, it appears that:
1. The default comparison is `Ordinal`, both in .NET Core and Fx.
2. .NET Framework does not contain any overload with a custom `StringComparison` value.
3. It looks like the analyzer does not have to warn, but probably does it for the sake of explicitness of the applied comparison. It's quite likely that many young developers will not be aware of this behavior under the hood.

Reply:
We provide the string comparison ["for clarity"](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1307). I guess because not everyone knows what is the default behavior. Using equals operator instead of `string.Equals(StringComparison.Ordinal)` makes sense because infix operators are inherently more readable, but when we have to use methods anyway the added clarify of an explicit string comparison is worth it.

Also the default behavior is not consistent across all string methods. While `string.Contains(value)` uses ordinal comparison by default, `string.StartsWith(value)` actually uses the current culture!
![image](https://user-images.githubusercontent.com/4007293/147604601-9eda9053-b1de-44c1-b8d8-0f6075f323ab.png)
I think it's worth keeping and using these methods (excluding `EqualsOrdinal`) just for peace of mind. They are only a little bit longer than the string member methods but they are properly consistent.

Reply:
I've just read about the different default `StringComparison` values for different string methods, too, and wondered why in the world they would do that.

With regard to just that, having the overloads is probably the safest bet, and it will still be shorter than adding the comparison value by hand.

Reply:
> Also the default behavior is not consistent across all string methods. While string.Contains(value) uses ordinal comparison by default, string.StartsWith(value) actually uses the current culture!

Ah, great. We do use [CA1310](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1310) as well, which specifies this only for such cases where it's not Ordinal by default. Which are [listed here](https://docs.microsoft.com/en-us/dotnet/standard/base-types/best-practices-strings#string-comparisons-that-use-the-current-culture).

So, I'd do this:
- Add *Ordinal() methods for those that don't use Ordinal comparison by default.
- Remove the other Ordinal methods but make it somehow apparent that we deliberately don't have them. I mentioned options above but I'm not sure. At least add docs. I'm on the fence here about the others except `EqualsOrdinal()` (that shouldn't be used) due to what Dávid says. In any case, if they remain, docs need to be added that you don't really have to use them.
- Disable CA1307 but keep CA1310.
- Roll back the changes in Hastlayer that thus are pointless.
- This does NOT affect the other extensions of ours that use IgnoreCase or InvariantCulture.

Reply:
- Isn't it going to be confusing to only have some ordinal methods?
- If we remove them, I vote for keeping a method stub and marking that Obsolete so it's very in-your-face which shouldn't be used. Nobody reads the docs. ;)
- Removed CA1307.

Reply:
Yeah, keeping all of them but marking the ones you actually shouldn't use with Obsolete is good.

Reply:
So as mentioned, please add *Ordinal() methods for those that don't use Ordinal comparison by default: https://docs.microsoft.com/en-us/dotnet/standard/base-types/best-practices-strings#string-comparisons-that-use-the-current-culture All of them, so we solve this for good.

How about also make the methods you shouldn't use throw an exception? Keep in mind that we only use these to make apparent that you should not use them.

Reply:
So that would be `IndexOf` and `LastIndexOf`. 
`Compare` and `CompareTo` don't have `StringComparison` overloads. Although a `CompareToOrdinal` extension method already exists in J2N. I don't think I've ever used these two anyway.

Reply:
Yep.

Reply:
I've found a `string.Compare` too, it's a static method. So those and the extension from J2N covers all eventualities after all.

Reply:
Great then.

Also note

> How about also make the methods you shouldn't use throw an exception? Keep in mind that we only use these to make apparent that you should not use them.

Reply:
Yes, I have already did that (https://github.com/Lombiq/Helpful-Libraries/pull/102/commits/e2309332c0171753a4d9ba7d61c9c54def60f63c)